test/lib/k8s_events.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lib
+
+import (
+	"context"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+// EventHandler is the callback type for the EventListener
+type EventHandler func(event *corev1.Event)
+
+// EventListener is a type that broadcasts new k8s events
+type EventListener struct {
+	cancel context.CancelFunc
+
+	lock     sync.Mutex
+	handlers []EventHandler
+}
+
+// NewEventListener creates a new event listener
+func NewEventListener(client kubernetes.Interface, namespace string, logf func(string, ...interface{})) *EventListener {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(
+		client,
+		0,
+		informers.WithNamespace(namespace),
+	)
+	eventsInformer := informerFactory.Core().V1().Events().Informer()
+
+	el := EventListener{
+		cancel: cancelCtx,
+	}
+
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			el.handle(obj.(*corev1.Event))
+		},
+	})
+
+	go func() {
+		eventsInformer.Run(ctx.Done())
+		logf("EventListener stopped")
+	}()
```

Comment:
In reconciler-test the event informer is injected into the context (something to keep in mind when moving this code over).

Reply:
Cool so we can avoid creating it manually!

Reply:
We'll still need this `EventListener`, right?

Reply:
Yup I think so.